testsuite/objects/__init__.py

```diff
@@ -1,12 +1,44 @@
 """Module containing base classes for common objects"""
 import abc
-from dataclasses import dataclass
+from dataclasses import dataclass, is_dataclass, fields
+from copy import deepcopy
 from functools import cached_property
-from typing import Literal, List
+from typing import Literal, List, Any
 
 from testsuite.objects.sections import Metadata, Identities, Authorizations, Responses
 
 
+def to_dict(obj):
+    """
+    This function converts dataclass object to dictionary.
+    While it works similar to `dataclasses.asdict` a notable change is usage of
+    overriding `to_dict()` function if dataclass contains it.
+    This function works recursively in lists, tuples and dicts. All other values are passed to copy.deepcopy function.
+    """
+    if hasattr(obj, "to_dict"):
+        return obj.to_dict()
+
+    result = {}
+    if is_dataclass(obj):
+        for field in fields(obj):
+            value = getattr(obj, field.name)
+            if value is None:
+                continue  # do not include None values
+
+            if is_dataclass(value):
+                result[field.name] = to_dict(value)
+            elif isinstance(value, (list, tuple)):
+                result[field.name] = type(value)(to_dict(i) for i in value)
+            elif isinstance(value, dict):
+                result[field.name] = type(value)((to_dict(k), to_dict(v)) for k, v in value.items())
+            else:
+                result[field.name] = to_dict(value)
+    else:
+        result = deepcopy(obj)
```

Comment:
I don't think that `to_dict` function should return an object deepcopy in any case. @mganisin WYT?

Reply:
As mentioned in other comment all this code is unnecessary in my opinion. Generally speaking this function resembles `dataclasses.asdict` therefore it should have same behavior. So `help(dataclasses.asdict)` would be authoritative answer if this should exist